srml/offences/src/lib.rs

```diff
@@ -0,0 +1,166 @@
+	}
+}
+
+decl_event!(
+	pub enum Event {
+		/// There is an offence reported of the given `kind` happened at the `session_index` and
+		/// (kind-specific) time slot. This event is not deposited for duplicate slashes.
+		Offence(Kind, SessionIndex, TimeSlot),
+	}
+);
+
+decl_module! {
+	/// Offences module, currently just responsible for taking offence reports.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+	}
+}
+
+impl<T: Trait, O: Offence<T::IdentificationTuple>> ReportOffence<T::AccountId, T::IdentificationTuple, O>
+	for Module<T> where
+	T::IdentificationTuple: Clone,
+{
+	fn report_offence(reporters: Vec<T::AccountId>, offence: O) {
+		let offenders = offence.offenders();
+		let time_slot = offence.time_slot();
+		let session = offence.session_index();
+		let validator_set_count = offence.validator_set_count();
+
+		// Check if an offence is already reported for the offender authorities
+		// and otherwise stores that report.
+		let mut new_offenders = BTreeSet::new();
+		let all_offenders = {
+			// We have to return the modified list of offending authorities. If we were to use
+			// `mutate` we would have to clone the whole list in order to return it.
+			let mut offending_authorities = <OffenceReports<T>>::get(&O::ID, &(session, time_slot));
+
+			for offender in offenders {
+				// TODO [slashing] This prevents slashing for multiple reports of the same kind at the same slot,
+				// note however that we might do that in the future if the reports are not exactly the same (dups).
+				// De-duplication of reports is tricky though, we need a canonical form of the report
+				// (for instance babe equivocation can have headers swapped).
+				if !offending_authorities.iter().any(|details| details.offender == offender) {
+					new_offenders.insert(offender.clone());
+					offending_authorities.push(OffenceDetails {
+						offender,
+						reporters: reporters.clone().into_iter().collect(),
+					});
+				}
+			}
+
+			offending_authorities
+		};
+
+		if new_offenders.is_empty() {
+			// The report contained only duplicates, so there is no need to slash again.
+			return
+		}
+
+		// We pushed new items in the offending_authorities, so update it.
+		<OffenceReports<T>>::insert(&O::ID, &(session, time_slot), &all_offenders);
+
+		let offenders_count = all_offenders.len() as u32;
+		let previous_offenders_count = offenders_count - new_offenders.len() as u32;
+
+		// The report is not a duplicate. Deposit an event.
+		Self::deposit_event(Event::Offence(O::ID, session, time_slot));
+
+		// The amount new offenders are slashed
+		let new_fraction = O::slash_fraction(offenders_count, validator_set_count);
+		// The amount previous offenders are slashed additionally.
+		//
+		// Since they were slashed in the past, we slash by:
+		// x = (new - prev) / (1 - prev)
+		// because:
+		// Y = X * (1 - prev)
+		// Z = Y * (1 - x)
+		// Z = X * (1 - new)
```

Comment:
This logic is all correct, but our main concern was whether per-bill can handle it. 
To keep the decimal representation from going out of control, we could always round the `slash_fraction` to the nearest 1000th or 10,000th. 

We will then incur some rounding error but are at much lower risk of overflow.